core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -133,12 +131,18 @@ private <T extends ContentFile<T>> PartitionSpec readPartitionSpec(InputFile inp
   private static <T extends ContentFile<T>> Map<String, String> readMetadata(InputFile inputFile) {
     Map<String, String> metadata;
     try {
-      try (AvroIterable<ManifestEntry<T>> headerReader =
-          Avro.read(inputFile)
+      try (CloseableIterable<ManifestEntry<T>> headerReader =
+          InternalData.read(FileFormat.AVRO, inputFile)
               .project(ManifestEntry.getSchema(Types.StructType.of()).select("status"))
               .classLoader(GenericManifestEntry.class.getClassLoader())
               .build()) {
-        metadata = headerReader.getMetadata();
+
+        if (headerReader instanceof InternalData.DataIterable) {
```

Comment:
This is a little bit of a workaround since we can't switch fully over to `DataIterable` due to binary incompatibility, but this is about the only place other than tests where it's used.

Reply:
I'm not sure that this is needed and I'd prefer not to add the interface if possible. We should not really be using the file metadata to recover things like the partition spec. I think that this is only used by older code paths that we didn't migrate to pass the id to spec map through.

At this point, I think we should go see where those methods are used and try to remove them, instead of adding this.

Reply:
There are currently a number of things that store in kv metadata, but InternalData isn't providing a way to expose that to consumers.  Right now this is limited to the [ManifestReader](https://github.com/apache/iceberg/blob/fa608705e1f7c1119eea6e0a60233a25820429ac/core/src/main/java/org/apache/iceberg/ManifestReader.java#L118), but more generally we don't have a way to access the kv data.

Reply:
I've removed the additional interface and we can handle the question of exposing metadata later.